### PR TITLE
feat!: Add AI online evaluations (Judge, Evaluator, IRunner)

### PR DESCRIPTION
**Pull request by mattrmc1**
BEGIN_COMMIT_OVERRIDE
feat: Add AI online evaluations (Judge, Evaluator, IRunner)
feat!: JudgeResult defaults for sampled and success changed from true to false per AIEVALS spec
END_COMMIT_OVERRIDE

## Summary

Adds **AI online evaluations** to `LaunchDarkly.ServerSdk.Ai`. A caller that supplies a `runnerFactory` to `LdAiClient` gets automatic judge evaluation wired into every `CompletionConfig` and `AgentConfig` — each returned config carries an `Evaluator` that can score model output against the judges declared in the flag's `judgeConfiguration`. When no `runnerFactory` is provided (or no judges are configured), configs receive a noop `Evaluator` so callers never need null checks.

Implements the [AIEVALS](https://github.com/launchdarkly/sdk-specs/tree/main/specs/AISDK-ai-sdk/AIEVALS-ai-online-evals) and [AIRUNNER](https://github.com/launchdarkly/sdk-specs/tree/main/specs/AISDK-ai-sdk/AIRUNNER-ai-runner) specs (sections 1.1–1.4). `createJudge` (AIEVALS 1.2) is intentionally omitted per .NET/Java convention — judges are created internally by the SDK, not by user code.

### New public types

```csharp
// Provider-facing runner interface (AIRUNNER 1.2)
public interface IRunner
{
    Task<RunnerResult> RunAsync(string input,
        IReadOnlyDictionary<string, object> outputType = null);
}

// Runner return type (AIRUNNER 1.3)
public sealed record RunnerResult(
    string Content,
    AiMetrics Metrics,
    object Raw = null,
    IReadOnlyDictionary<string, object> Parsed = null);

// Evaluation orchestrator (AIEVALS 1.4)
public sealed class Evaluator
{
    public static Evaluator Noop();
    public Task<IReadOnlyList<JudgeResult>> EvaluateAsync(string input, string output);
}

// Single-judge executor (AIEVALS 1.1)
public sealed class Judge
{
    public LdAiJudgeConfig Config { get; }
    public IRunner Runner { get; }
    public Task<JudgeResult> EvaluateAsync(string input, string output, double? samplingRate = null);
    public Task<JudgeResult> EvaluateMessagesAsync(
        IReadOnlyList<LdAiConfigTypes.Message> messages,
        RunnerResult runnerResult, double? samplingRate = null);
}
```

### `LdAiClient` changes

```csharp
// New optional parameter
public LdAiClient(ILaunchDarklyClient client,
    Func<LdAiJudgeConfig, IRunner> runnerFactory = null);
```

When `runnerFactory` is non-null, `ConfigFactory.BuildEvaluator` iterates the flag's `judgeConfiguration`, evaluates each judge key as a flag variation, creates a `Judge` + `IRunner` pair per enabled judge, and attaches the resulting `Evaluator` to the config. Disabled judges, null runners, and initialization exceptions are logged and skipped — no single judge failure prevents the others from being built.

### `LdAiConfig` base class

All config types (`LdAiCompletionConfig`, `LdAiAgentConfig`, `LdAiJudgeConfig`) now carry an `Evaluator` property via the base class. `LdAiJudgeConfig` always receives `Evaluator.Noop()` (judges don't evaluate themselves).

### `JudgeResult` changes

Updated to match AIEVALS 1.3.1 defaults:

| Field | Before | After |
|-------|--------|-------|
| `MetricKey` | required | optional (default `null`) |
| `Score` | required | optional (default `0.0`) |
| `Sampled` | default `true` | default `false` |
| `Success` | default `true` | default `false` |
| `ErrorMessage` | — | new field |
| `Reasoning` | — | new field |

### Null safety (ref: [java-core#175 discussion](https://github.com/launchdarkly/java-core/pull/175#discussion_r3498911230))

`BuildEvaluator` handles every failure mode without throwing:
- `runnerFactory == null` or empty `judgeConfiguration` → `Evaluator.Noop()`
- Runner factory returns `null` → warn + skip judge
- Judge config disabled → warn + skip judge
- Any exception during judge init → warn + skip judge
- Missing `evaluationMetricKey` → `JudgeResult(success: false, errorMessage: ...)`
- Score out of `[0, 1]` range → `JudgeResult(success: false, errorMessage: ...)`
- Sampling rate `NaN`/`Infinity`/negative/`> 1.0` → normalized to safe bounds

`Judge` constructor still uses `ArgumentNullException` guards, but these are never hit in practice because `BuildEvaluator` validates inputs before construction.

## Migration

**None required.** The `LdAiClient` constructor gains an optional `runnerFactory` parameter (default `null`) — existing callers are unaffected. `JudgeResult` default changes are source-compatible (all parameters are now optional with safe defaults). No members removed or renamed.

## Test plan

- [ ] `dotnet test pkgs/sdk/server-ai/test/LaunchDarkly.ServerSdk.Ai.Tests.csproj --framework net8.0` passes
- [ ] `JudgeTest` (435 lines) covers: successful evaluation with score/reasoning extraction, sampling skip path, `samplingRate` edge cases (`NaN`, negative, `> 1.0`), runner exception handling, missing `evaluationMetricKey` validation, out-of-range score rejection with `errorMessage`, `evaluateMessages` formatting, null/empty message handling
- [ ] `EvaluatorTest` (210 lines) covers: noop returns empty list, multi-judge execution with per-judge sampling, missing judge key logs warning and skips, noop does not log warnings
- [ ] `LdAiCompletionConfigTest` (189 lines added) covers: `Evaluator` attached when `runnerFactory` provided, noop `Evaluator` when no runner factory, noop when `judgeConfiguration` is empty, disabled judge skipped, null runner skipped
- [ ] `LdAiJudgeConfigTest` covers: `JudgeResult` `ErrorMessage` and `Reasoning` fields, judge config always receives noop `Evaluator`
- [ ] `LdAiConfigTrackerTest` covers: `TrackJudgeResult` with new optional fields

---

> [!NOTE]
> **Medium Risk**
> New optional API path triggers extra model calls for judges and changes `JudgeResult` default semantics, which can affect tracking behavior for existing callers.
> 
> **Overview**
> Adds **AI online evaluations** to the server AI SDK: callers can pass an optional `runnerFactory` into `LdAiClient`, and `ConfigFactory` builds a real `Evaluator` from each flag’s `judgeConfiguration` (fetching judge configs, skipping disabled/null/failed judges) or `Evaluator.Noop()` when no factory or no judges.
> 
> **New public surface:** `IRunner`, `RunnerResult`, `Judge` (model call with structured score/reasoning schema, sampling, validation), and `Evaluator` (`EvaluateAsync` over configured judges; does not call `TrackJudgeResult`). Completion and agent configs (including default/fallback paths and graph agents) expose a non-null `Evaluator` on `LdAiConfig`; judge configs always get noop.
> 
> **Config/parsing tweaks:** `JudgeConfiguration.Judge.SamplingRate` is `double?` (omit/null → evaluate 100%); defaults only serialize `samplingRate` when set. Judge prompt variables from the parent config are forwarded when building nested judge configs.
> 
> **Breaking (`JudgeResult`):** defaults for `Sampled` and `Success` flip to `false`; `MetricKey`/`Score` optional; new `ErrorMessage` and `Reasoning` fields per AIEVALS.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2f97f4249269ffccd78b9f98fa627d86e517c2b5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkgs/sdk/server-ai/src/Evals/Evaluator.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            if (!_judges.TryGetValue(judgeEntry.Key, out var judge))
+            {
+                _logger?.Warn("Evaluator: judge '{0}' not found; skipping", judgeEntry.Key);
+                continue;
+            }
+
+            var result = await judge.EvaluateAsync(input, output, judgeEntry.SamplingRate);
+            results.Add(result);
+        }
```
[the remark this answers is not in the document]
Comment (mattrmc1) [reply pkgs/sdk/server-ai/src/Evals/Evaluator.cs:90]: Is this actually a bug? Should the SDK be defensive about this or faithfully evaluate the list it receives (even if duplicates are present)?

cc @jsonbailey

File: pkgs/sdk/server-ai/src/Config/ConfigFactory.cs [hunk 2]
```diff
@@ -247,6 +265,49 @@ private LdAiJudgeConfig BuildJudgeFromDefault(
...
+        }
+
+        var judges = new Dictionary<string, Judge>();
+        foreach (var judgeEntry in judgeConfiguration.Judges)
+        {
+            try
+            {
+                var defaultValue = LdAiJudgeConfigDefault.Disabled;
+                var ldValue = _client.JsonVariation(judgeEntry.Key, context, defaultValue.ToLdValue());
+                var judgeConfig = BuildJudgeConfig(judgeEntry.Key, ldValue, context, defaultValue, variables, graphKey: graphKey);
```
[the remark this answers is not in the document]
Comment (mattrmc1) [reply pkgs/sdk/server-ai/src/Config/ConfigFactory.cs:283]: Not a bug. Attached judges are operational -- they execute against a model provider via IRunner.RunAsync and require interpolated prompts to function. A judge receiving raw {{variable}} tokens would produce nonsense. When fetching a judge directly with `interpolation: false`, the raw {{variable}} tokens will return correctly

[the remark this answers is not in the document]
Comment (mattrmc1) [reply pkgs/sdk/server-ai/src/Config/ConfigFactory.cs:283]: Messages when fetching toxicity judge as template:

```json
[
    {
        "content": "You are a business information accuracy and safety expert. Evaluate....",
        "role": "system"
    },
    {
        "content": "MESSAGE HISTORY:\n{{message_history}}\n",
        "role": "assistant"
    },
    {
        "content": "RESPONSE TO EVALUATE:\n{{response_to_evaluate}}\n",
        "role": "user"
    }
]
```

File: pkgs/sdk/server-ai/src/Evals/Evaluator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsonbailey) [reply pkgs/sdk/server-ai/src/Evals/Evaluator.cs:90]: If there are duplicates, it would be a bug in the server as it should never allow duplicates. I don't think there is any harm in being defensive about it, but I also don't think its really necessary. I would agree this is low severity and may not be worth the extra check.

**Comment by mattrmc1** [comment]
Converting to draft for now

**Review by tanderson-ld — CHANGES_REQUESTED** [review]
Not reviewing yet in draft.
